tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc

```diff
@@ -0,0 +1,176 @@
+    }
+    if (idx >= 0) {
+      if (i != idx) {
+        return errors::Internal(
+            "Profile index of engine config is different from resource profile "
+            "index: ",
+            i, " != ", idx);
+      }
+      VLOG(1) << "Added optimization profile " << profiles_[i].DebugString()
+              << " to builder config.";
+    } else {
+      VLOG(ERROR) << "Failed to add optimization profile "
+                  << profiles_[i].DebugString()
+                  << ". This usually happens when profile is invalid.";
+    }
+  }
+  if (config->getNbOptimizationProfiles() == 0) {
+    return errors::Internal("Failure in adding an optimization profile.");
+  }
+  // if TRT_VERSION < 6, then we do not need to add
+  return Status::OK();
+}
+#endif
+
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+Status TrtShapeOptimizationProfile::ConfigureBuilder(
+    nvinfer1::IBuilder* builder, nvinfer1::IBuilderConfig* config,
+    const nvinfer1::INetworkDefinition* network) {
+  AddProfiles(builder, config, network);
+  return Status::OK();
+}
+#endif
+
+int TrtShapeOptimizationProfile::GetProfileNumber(
+    std::vector<TensorShape> shapes) {
+  for (int i = 0; i < profiles_.size(); i++) {
+    if (profiles_[i].IncludesShapes(shapes)) {
+      return i;
+    }
+  }
+  VLOG(1) << "Profile not found for input shapes " << DebugString(shapes)
+          << ".";
+  return -1;
+}
+
+Status TrtShapeOptimizationProfile::CreateExecutionContexts(
+    nvinfer1::ICudaEngine* engine,
+    std::vector<TrtUniquePtrType<nvinfer1::IExecutionContext>>& exec_context) {
+  int i = 0;
+  // The following loop runs once if we have static shapes, to create a single
+  // execution context without profiles. In dynamic mode we create one context
+  // for each profile and set the corresponding optimization profile.
+  do {
+    VLOG(1) << "Creating execution context " << i;
+    nvinfer1::IExecutionContext* ctx = engine->createExecutionContext();
+    if (ctx == nullptr) {
+      return errors::Internal("Failed to create execution context");
+    }
+    if (i > 0) {
+      // This condition is needed for two reasons:
+      // - using static shapes we do not have any profiles so we cannot call
+      //   set optimizationprofiles.
+      // - The 0th profile is set implicitly for the first execution context
+      //   therefore we do not need to set.
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+      bool stat = ctx->setOptimizationProfile(i);
+      if (!stat) {
+        ctx->destroy();
+        return errors::Internal("Could not set TRT optimization profile.");
+      }
+#endif
+    }
+    exec_context.push_back(
+        std::move(TrtUniquePtrType<nvinfer1::IExecutionContext>(ctx)));
+    i++;
+  } while (i < profiles_.size());
+
+  return Status::OK();
+}
+
+Status TrtShapeOptimizationProfile::RestoreProfiles(
+    const nvinfer1::ICudaEngine* engine) {
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+  if (!engine || engine->hasImplicitBatchDimension()) {
```

Comment:
I see this compilation error no member named 'hasImplicitBatchDimension' in 'nvinfer1::ICudaEngine'